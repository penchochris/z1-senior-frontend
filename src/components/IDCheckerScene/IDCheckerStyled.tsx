import styled from 'styled-components';

interface OverlayProps {
  isRejected?: boolean;
}

interface CameraWrapperProps extends OverlayProps {
  imageSource?: string;
}

export const Title = styled.h2`
  height: 24px;
  width: 282px;
  color: #000000;
  font-family: Roboto;
  font-size: 21px;
  font-weight: bold;
  line-height: 24px;
  text-align: center;
  left: 50%;
`;

export const Text = styled.p`
  height: 48px;
  width: 314px;
  color: #000000;
  font-family: Roboto;
  font-size: 16px;
  line-height: 24px;
  text-align: center;
  padding-bottom: 27px;
`;

export const OverlayLabel = styled.p`
	height: 12px;
	width: 67px;
	color: #FFFFFF;
	font-family: Roboto;
	font-size: 12px;
	font-weight: bold;
	letter-spacing: 0.75px;
	line-height: 12px;
  text-align: right;
`

export const CameraWrapper = styled.div`
  display: flex;
  flex-flow: column;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  height: 179px;
  width: 289px;
  border: ${(props:CameraWrapperProps) => 
    props.isRejected 
      ? '2px solid #C00000'
      : '2px solid #69CC8B'
  };
  border-radius: 18px;
  background-image: ${(props:CameraWrapperProps) => 
    `url(${props.imageSource})`
  };
  background-repeat: no-repeat;
  background-size: 100% 100%;
  position: relative;
  overflow: hidden;
`;

export const Overlay = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
	height: 30px;
	width: 105px;
	border-radius: 4px;
  background-color: ${(props:OverlayProps) => props.isRejected ? '#C00000' : '#69CC8B'};
  align-items: center;
  position: absolute;
  bottom: -15px;
  right: 30px;
`;

export const OverlayWrapper = styled.div`
  position: relative;
`;



