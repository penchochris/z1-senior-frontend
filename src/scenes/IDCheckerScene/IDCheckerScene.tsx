import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import axios from 'axios';

import Header from '../../components/common/Header/Header';
import { APPROVED } from '../../consts';
import ScanID from '../../components/IDCheckerScene/ScanID/ScanID';
import TakePictureID from '../../components/IDCheckerScene/TakePictureID/TakePictureID';

const Wrapper = styled.div`
  display: flex;
  flex-flow: column;
  height: 100vh;
  background-color: ${(props:isWebcamOn) => 
    props.isWebcamOn && 'rgba(20,28,38,0.75)'
  };
`;

const WrapperMain = styled.div`
  display: flex;
  flex-flow: column;
  align-items: center;
  height: 100vh;
  width: 100vw;
  padding-top: 32px;
`;

interface isWebcamOn {
  isWebcamOn: boolean;
}

interface ImageStatusResponse {
  summary: {
    outcome: string;
  };
}

const ScanCardScene: React.FC = () => {

  const [ isWebcamOn, setIsWebcamOn ] = useState<boolean>();
  const [ isRejected, setIsRejected ] = useState<boolean>();
  const [ imageSource, setImageSource ] = useState<string>();
  
  const webcamRef = React.useRef<any>(null);

  useEffect(() => {
    let timer: any;
    if(isWebcamOn) {
      timer = setTimeout( async () => {
        const payload = webcamRef.current.getScreenshot();
        //Here we should be supposed to send a payload :)
        const { data } = await axios.post<ImageStatusResponse>('https://front-exercise.z1.digital/evaluations');
        
        data.summary && data.summary.outcome === APPROVED
          ? setIsRejected(false)
          : setIsRejected(true);
        
        setImageSource(payload)
        setIsWebcamOn(false);
      }, 3000);
    }
    return () => clearTimeout(timer);
  }, [isWebcamOn, imageSource, setIsWebcamOn, setImageSource, webcamRef]);

  const handleOnClick = () => {
    setImageSource('');
    setIsWebcamOn(true);
  }

  const handleCancel = () => {
    setImageSource('');
    setIsWebcamOn(false);
    setIsRejected(undefined);
  }

  return (
    <Wrapper isWebcamOn={isWebcamOn}>
      <Header label={ !isWebcamOn || imageSource ? 'BankClient' : '' }/>
      <WrapperMain>
      {
        isWebcamOn
          ? <ScanID webcamRef={webcamRef} handleCancel={handleCancel}/>
          : <TakePictureID
              imageSource={imageSource}
              handleOnClick={handleOnClick}
              isRejected={isRejected}
            />
            
      }
      </WrapperMain>
    </Wrapper>
  );
}

export default ScanCardScene;