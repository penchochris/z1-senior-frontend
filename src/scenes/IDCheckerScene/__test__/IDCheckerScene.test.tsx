import React, { useState } from 'react';
import ReactDOM from 'react-dom';

import { StateMock } from '@react-mock/state';
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer';

import IDCheckerScene from '../IDCheckerScene';

interface IDCheckerSceneProps {
  isWebcamOn: boolean;
  isRejected: boolean;
  imageSource: string;
}

const mockIDCheckerSceneState = ({ isWebcamOn, isRejected, imageSource }: IDCheckerSceneProps) =>
  <StateMock state={{ isWebcamOn, isRejected, imageSource }}>
    <IDCheckerScene />
  </StateMock>


it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<IDCheckerScene/>, div);
})

it('matches snapshot', () => {
  const tree = renderer.create(<IDCheckerScene/>).toJSON();
  expect(tree).toMatchSnapshot();
})

it('matches snapshot when click TAKE PICTURE button', () => {
  const tree = renderer.create(mockIDCheckerSceneState({ isWebcamOn: true, isRejected: undefined, imageSource: '' })).toJSON();
  expect(tree).toMatchSnapshot();
})

it('matches snapshot when rejects picture', () => {
  const tree = renderer.create(mockIDCheckerSceneState({ isWebcamOn: false, isRejected: true, imageSource: 'image' })).toJSON();
  expect(tree).toMatchSnapshot();
})

it('matches snapshot when accepts picture', () => {
  const tree = renderer.create(mockIDCheckerSceneState({ isWebcamOn: false, isRejected: false, imageSource: 'image' })).toJSON();
  expect(tree).toMatchSnapshot();
})