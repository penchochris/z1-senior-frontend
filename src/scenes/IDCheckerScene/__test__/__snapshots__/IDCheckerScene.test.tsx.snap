// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="sc-fznyAO jWQUuP"
>
  <div
    className="sc-AxjAm gYslqW"
  >
    <h1
      className="sc-AxirZ fCAmhS"
    >
      BankClient
    </h1>
  </div>
  <div
    className="sc-fznKkj cnDexW"
  >
    <h2
      className="sc-AxiKw jXErWB"
    >
      Scan your ID
    </h2>
    <p
      className="sc-AxhCb jIBAms"
    >
      Take a picture. It may take time to validate your personal information
    </p>
    <div
      className="sc-fzpans fzRXyq"
    >
      <div
        className="sc-fzplWN iUmTbq"
      >
        <div
          className="sc-fzozJi fIvUiB"
          onClick={[Function]}
        >
          <p
            className="sc-fzoLsD kyhDmr"
          >
            TAKE PICTURE
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`matches snapshot when accepts picture 1`] = `
<div
  className="sc-fznyAO jWQUuP"
>
  <div
    className="sc-AxjAm gYslqW"
  >
    <h1
      className="sc-AxirZ fCAmhS"
    >
      BankClient
    </h1>
  </div>
  <div
    className="sc-fznKkj cnDexW"
  >
    <h2
      className="sc-AxiKw jXErWB"
    >
      Scan your ID
    </h2>
    <p
      className="sc-AxhCb jIBAms"
    >
      Take a picture. It may take time to validate your personal information
    </p>
    <div
      className="sc-fzpans fzRXyq"
    >
      <div
        className="sc-fzplWN iUmTbq"
      >
        <div
          className="sc-fzozJi fIvUiB"
          onClick={[Function]}
        >
          <p
            className="sc-fzoLsD kyhDmr"
          >
            TAKE PICTURE
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`matches snapshot when click TAKE PICTURE button 1`] = `
<div
  className="sc-fznyAO jWQUuP"
>
  <div
    className="sc-AxjAm gYslqW"
  >
    <h1
      className="sc-AxirZ fCAmhS"
    >
      BankClient
    </h1>
  </div>
  <div
    className="sc-fznKkj cnDexW"
  >
    <h2
      className="sc-AxiKw jXErWB"
    >
      Scan your ID
    </h2>
    <p
      className="sc-AxhCb jIBAms"
    >
      Take a picture. It may take time to validate your personal information
    </p>
    <div
      className="sc-fzpans fzRXyq"
    >
      <div
        className="sc-fzplWN iUmTbq"
      >
        <div
          className="sc-fzozJi fIvUiB"
          onClick={[Function]}
        >
          <p
            className="sc-fzoLsD kyhDmr"
          >
            TAKE PICTURE
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`matches snapshot when rejects picture 1`] = `
<div
  className="sc-fznyAO jWQUuP"
>
  <div
    className="sc-AxjAm gYslqW"
  >
    <h1
      className="sc-AxirZ fCAmhS"
    >
      BankClient
    </h1>
  </div>
  <div
    className="sc-fznKkj cnDexW"
  >
    <h2
      className="sc-AxiKw jXErWB"
    >
      Scan your ID
    </h2>
    <p
      className="sc-AxhCb jIBAms"
    >
      Take a picture. It may take time to validate your personal information
    </p>
    <div
      className="sc-fzpans fzRXyq"
    >
      <div
        className="sc-fzplWN iUmTbq"
      >
        <div
          className="sc-fzozJi fIvUiB"
          onClick={[Function]}
        >
          <p
            className="sc-fzoLsD kyhDmr"
          >
            TAKE PICTURE
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
